### PR TITLE
Initial changes for GL2 Residency.

### PR DESCRIPTION
**Pull request by kjayapra-amd**
## Motivation

Adds GL2 Residency Control (Persisting L2 Cache) to allow applications to reserve a portion of the GPU's L2 cache for persistent data. This improves performance for ML inference workloads with repeated access to small weight arrays by reducing memory bandwidth pressure.

## Technical Details

Added `hipDeviceProp_t.persistingL2CacheMaxSize` property and `hipLimitPersistingL2CacheSize` limit type for use with `hipDeviceSetLimit()`/`hipDeviceGetLimit()`. Implemented using HSA runtime APIs (`hsa_amd_agent_set_attribute`) with dynamic symbol loading support added to ROCclr.

## Issue Tracking
AIRUNTIME-2

## Test Plan

Unit tests cover API correctness, boundary conditions, functional validation, and concurrency. Performance tests measure scaling at different L2 reservation levels (0%, 25%, 50%, 75%, 100%).

## Test Result

Unit tests cover API correctness, boundary conditions, functional validation, and concurrency. Performance tests measure scaling at different L2 reservation levels (0%, 25%, 50%, 75%, 100%).

## Submission Checklist

- [X] Look over the contributing guidelines at https://github.com/ROCm/rocm-systems/blob/develop/CONTRIBUTING.md.

### CODE REVIEW

File: projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+    for (int w = 0; w < numWeights; w++) {
+      result += weights[w] * inputs[idx * numWeights + w];
+    }
+  }
+
+  outputs[idx] = result;
+}
+
+bool isGL2ResidencySupported() {
+#if __linux__
```
Comment (AlexeySachkov) [inline projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc]: If the API is generic, why do we need any preprocessor tricks? By always performing the query we also test that it works on Windows (by works I mean correctly reports 0 without errors or crashes).

File: projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+  HIP_CHECK(hipGetDeviceProperties(&props, 0));
+
+  size_t maxL2 = props.persistingL2CacheMaxSize;
+
+  std::cout << "\n========================================" << std::endl;
+  std::cout << "GL2 Residency Scaling Performance Test" << std::endl;
+  std::cout << "========================================" << std::endl;
+  std::cout << "Device: " << props.gcnArchName << std::endl;
+  std::cout << "Max L2 Reservation: " << maxL2 / 1024 / 1024 << " MB" << std::endl;
+  std::cout << "========================================\n" << std::endl;
```
Comment (AlexeySachkov) [inline projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc]: We should probably use built-in `catch2` macro like [`CAPTURE`](https://github.com/catchorg/Catch2/blob/devel/docs/logging.md#quickly-capture-value-of-variables-or-expressions)

File: projects/clr/rocclr/device/rocm/rocdevice.cpp [hunk 3]
```diff
@@ -3139,6 +3139,44 @@ bool Device::UpdateScratchLimitCurrent(size_t limit) const {
...
+                         &persistingL2CacheSizeMax);
+  if (HSA_STATUS_SUCCESS != ret) {
+    LogPrintfError("HSA_AMD_AGENT_INFO_MAX_PERSISTING_L2_CACHE_SIZE cannot be queried! Err: 0x%xh", ret);
+    return 0;
+  }
+  return static_cast<size_t>(persistingL2CacheSizeMax);
+};
+
+// ================================================================================================
+bool Device::UpdatePersistingL2CacheSizeCurrent(size_t size) const {
```
Comment (AlexeySachkov) [inline projects/clr/rocclr/device/rocm/rocdevice.cpp:3185]: I understand that the user-facing API uses `size_t` simply because it is a single entry point to modify every possible value. But why do we need internal APIs to match it? Wouldn't it be better for all internal APIs to be consistent and match the actual data type used for that info and leave conversion to the user-facing API?

File: projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc (hunk 2 above)
Comment (kjayapra-amd) [reply projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc]: Yes, this is a oversight. Needs clean up. Tests are still WIP.

File: projects/clr/rocclr/device/rocm/rocdevice.cpp (hunk 3 above)
Comment (kjayapra-amd) [reply projects/clr/rocclr/device/rocm/rocdevice.cpp:3185]: The device class member function interface should be backend agonostic. For ex, if PAL backend exposes their API as int32_t, then that would be a conversion from size_t --> uint32_t --> int32_t.

Keeping the argument type same as HIP interface makes sense in this case and conversion closer to HSA API shows the reason for type conversion.

File: projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc (hunk 1 above)
Comment (kjayapra-amd) [reply projects/hip-tests/catch/performance/scenarios/residency/gl2_scaling.cc]: Tests are still WIP. Needs clean up.

File: projects/clr/hipamd/include/hip/amd_detail/hip_prof_str.h [hunk 4]
```diff
@@ -7579,6 +7579,11 @@ static inline void hipApiArgsInit(hip_api_id_t id, hip_api_data_t* data) {
     case HIP_API_ID_hipDeviceGetLimit:
       if (data->args.hipDeviceGetLimit.pValue) data->args.hipDeviceGetLimit.pValue__val = *(data->args.hipDeviceGetLimit.pValue);
       break;
+// hipDeviceGetLuid[('char*', 'luid'), ('unsigned int*', 'deviceNodeMask'), ('hipDevice_t', 'device')]
+    case HIP_API_ID_hipDeviceGetLuid:
+      data->args.hipDeviceGetLuid.luid = (data->args.hipDeviceGetLuid.luid) ? strdup(data->args.hipDeviceGetLuid.luid) : NULL;
+      if (data->args.hipDeviceGetLuid.deviceNodeMask) data->args.hipDeviceGetLuid.deviceNodeMask__val = *(data->args.hipDeviceGetLuid.deviceNodeMask);
+      break;
```
Comment (mangupta) [inline projects/clr/hipamd/include/hip/amd_detail/hip_prof_str.h]: 🤖 AI-assisted pre-review (for @mangupta)

**[BLOCKING]** hipApiArgsInit now strdup()s the hipDeviceGetLuid output buffer (uninitialized over-read, leak, and luid__val no longer populated)

The regenerated `HIP_API_ID_hipDeviceGetLuid` case in `hipApiArgsInit` changed from copying one byte (`luid__val = *(luid)`) to `data->args.hipDeviceGetLuid.luid = strdup(data->args.hipDeviceGetLuid.luid)`. `luid` is an *output* parameter (an 8-byte caller buffer that the API is about to fill), not a NUL-terminated input string, so `strdup` scans uninitialized caller memory for a NUL byte, allocates a copy that is never freed, and overwrites the recorded caller pointer with that copy.

**Suggested fix:** Restore the non-string handling for this output parameter (`if (luid) luid__val = *(luid);`, or capture the fixed 8-byte LUID explicitly) and do not `strdup` an output buffer. If the file is machine-generated, fix the generator's type classification so `char*` output parameters are not treated as input C strings, and regenerate.

File: projects/hip-tests/catch/unit/device/hipDevicePersistingL2Cache.cc [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+ */
+HIP_TEST_CASE(Unit_hipDeviceLimit_PersistingL2Cache_NegativeParams) {
+  SECTION("nullptr parameter") {
+    HIP_CHECK_ERROR(hipDeviceGetLimit(nullptr, hipLimitPersistingL2CacheSize),
+                    hipErrorInvalidValue);
+  }
+
+  SECTION("Invalid limit value") {
+    size_t val;
+    HIP_CHECK_ERROR(hipDeviceGetLimit(&val, hipLimitRange), hipErrorInvalidValue);
```
Comment (mangupta) [inline projects/hip-tests/catch/unit/device/hipDevicePersistingL2Cache.cc]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** NegativeParams test expects hipErrorInvalidValue for an unsupported limit, but hipDeviceGetLimit returns hipErrorUnsupportedLimit

`SECTION("Invalid limit value")` asserts `HIP_CHECK_ERROR(hipDeviceGetLimit(&val, hipLimitRange), hipErrorInvalidValue)`, but the `default:` arm of `hipDeviceGetLimit` returns `hipErrorUnsupportedLimit` for any unrecognized limit. The assertion will fail on every device.

**Suggested fix:** Expect `hipErrorUnsupportedLimit` for `hipLimitRange` (keep `hipErrorInvalidValue` only for the nullptr case), and add `(void)hipGetLastError();` after each intentionally failing call.

File: projects/hip/include/hip/hip_runtime_api.h [hunk 6]
```diff
@@ -820,6 +820,11 @@ enum hipLimit_t {
                                        ///< device. Must be between hipExtLimitScratchMin and
                                        ///< hipExtLimitScratchMaxValid values. Valid only on Rocm
                                        ///< device. This can be modified.
+  hipLimitPersistingL2CacheSize = 0x1003, ///< Requested persisting L2 cache size in bytes for the
+                                        ///< current device. This size can be set to a value up to
+                                        ///< persistingL2CacheMaxSize reported in hipDeviceProp_t.
+                                        ///< Only supported on MI400/MI450 and later. This can be
+                                        ///< modified.
   hipLimitRange                        ///< Supported limit range
```
Comment (mangupta) [inline projects/hip/include/hip/hip_runtime_api.h]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Inserting hipLimitPersistingL2CacheSize = 0x1003 before hipLimitRange changes hipLimitRange's numeric value

The new enumerator takes the value `0x1003`, which was previously the implicit value of the trailing sentinel `hipLimitRange`. `hipLimitRange` is a public enumerator and silently becomes `0x1004`, so the numeric meaning of `0x1003` changes between header versions.

**Suggested fix:** Give the new value a number above the current sentinel (or append it after the last real limit and keep `hipLimitRange` pinned to an explicit value that does not move), so no previously-valid enumerator value is reassigned.

File: projects/clr/hipamd/src/hip_device_runtime.cpp [hunk 7]
```diff
@@ -546,6 +546,9 @@ hipError_t hipDeviceGetLimit(size_t* pValue, hipLimit_t limit) {
     case hipExtLimitScratchCurrent:
       *pValue = hip::getCurrentDevice()->devices()[0]->ScratchLimitCurrent();
       break;
+    case hipLimitPersistingL2CacheSize:
+      *pValue = hip::getCurrentDevice()->devices()[0]->PersistingL2CacheSizeCurrent();
+      break;
```
Comment (mangupta) [inline projects/clr/hipamd/src/hip_device_runtime.cpp]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Unsupported/failed HSA query is reported as hipSuccess with a value of 0 for hipLimitPersistingL2CacheSize

`PersistingL2CacheSizeCurrent()` returns 0 both when the feature is genuinely set to 0 and when the HSA query fails (unsupported attribute, older ROCr). `hipDeviceGetLimit` copies that 0 out and returns `hipSuccess`, so an application cannot distinguish "feature unavailable" from "reservation currently zero". The same swallow exists on the set path: on a device where `PersistingL2CacheSizeMax()` fails, `UpdatePersistingL2CacheSizeCurrent(0)` still calls `hsa_amd_agent_set_attribute`, and on non-ROCm backends the base-class default (`device.hpp`, new `UpdatePersistingL2CacheSizeCurrent` returning `true`) makes `hipDeviceSetLimit` report success while `hipDeviceGetLimit` keeps returning 0.

**Suggested fix:** Propagate the backend failure: have the rocclr accessors report success/failure separately from the value (e.g. out-parameter + bool), and return `hipErrorUnsupportedLimit` from `hipDeviceGetLimit`/`hipDeviceSetLimit` when the attribute is not queryable; make the `device.hpp` default `UpdatePersistingL2CacheSizeCurrent` return `false` so non-ROCm backends do not report success for a no-op.

File: projects/clr/rocclr/device/rocm/rocdevice.cpp [hunk 8]
```diff
@@ -3155,6 +3155,51 @@ bool Device::UpdateScratchLimitCurrent(size_t limit) const {
...
+  const size_t max_size = PersistingL2CacheSizeMax();
+  if (size > max_size) {
+    LogPrintfError("Persisting L2 cache size %zu exceeds device max %zu", size, max_size);
+    return false;
+  }
+  // HSA API expects size_t* (not uint32_t*) per hsa_ext_amd.h documentation
+  // Make a mutable copy since HSA API takes non-const void*
+  size_t cache_size = size;
+  hsa_status_t ret = Hsa::agent_set_attribute(bkendDevice_,
+      HSA_AMD_AGENT_ATTRIBUTE_REQUEST_PERSISTING_L2_CACHE_SIZE, &cache_size);
```
Comment (mangupta) [inline projects/clr/rocclr/device/rocm/rocdevice.cpp]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Same HSA persisting-L2 attribute is read as uint32_t but written through a size_t buffer

`PersistingL2CacheSizeCurrent()` passes `&persistingL2CacheSize` where the variable is `uint32_t`, while `UpdatePersistingL2CacheSizeCurrent()` passes `&cache_size` where the variable is `size_t` (8 bytes). Only one width can match what ROCr reads/writes for this attribute; the other silently reads or writes the wrong number of bytes.

**Suggested fix:** Confirm the declared payload type of the attribute in `hsa_ext_amd.h` and use that exact type on both the get and set paths, converting to/from `size_t` only at the rocclr boundary; drop the comment that asserts a different width than the getter uses.

File: projects/clr/hipamd/src/hip_device.cpp [hunk 9]
```diff
@@ -770,7 +770,7 @@ hipError_t ihipGetDeviceProperties(hipDeviceProp_tR0600* props, int device) {
   // Caching behavior
   deviceProps.globalL1CacheSupported = 1;
   deviceProps.localL1CacheSupported = 1;
-  deviceProps.persistingL2CacheMaxSize = info.l2CacheSize_;
+  deviceProps.persistingL2CacheMaxSize = deviceHandle->PersistingL2CacheSizeMax();
```
Comment (mangupta) [inline projects/clr/hipamd/src/hip_device.cpp]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** persistingL2CacheMaxSize silently changes from l2CacheSize to 0 on every currently shipping device, with no doc or changelog update

`hipDeviceProp_t::persistingL2CacheMaxSize` was reported as `info.l2CacheSize_`; it now comes from a live HSA query that returns 0 on every device/ROCr that does not implement the new attribute. This is an observable change to a public device property, and it is also a new HSA round-trip on each `hipGetDeviceProperties` call instead of a cached `info_` field like the neighbouring properties.

**Suggested fix:** Document the new semantics (public doxygen for the property and/or `projects/clr/CHANGELOG.md`) alongside the new `hipLimitPersistingL2CacheSize` enumerator, and cache the max value in the device `info_` structure at init time instead of issuing an HSA query on every `hipGetDeviceProperties` call.

File: projects/clr/rocclr/device/devkernel.cpp [hunk 10]
```diff
@@ -48,10 +48,17 @@ amd_comgr_status_t getMetaBuf(const amd_comgr_metadata_node_t meta,
 bool getValueFromIsaMeta(const std::string& isa, const char* key, std::string& retValue) {
   amd_comgr_metadata_node_t isaMeta;
 
-  amd_comgr_status_t status = amd::Comgr::get_isa_metadata(isa.c_str(), &isaMeta);
+  // Workaround: Strip -strict suffix for gfx1250 as metadata may not exist for strict variant
+  std::string lookup_isa = isa;
+  size_t strict_pos = lookup_isa.find("-strict");
+  if (strict_pos != std::string::npos) {
+    lookup_isa = lookup_isa.substr(0, strict_pos);
+  }
```
Comment (mangupta) [inline projects/clr/rocclr/device/devkernel.cpp]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** Unrelated gfx1250 "-strict" ISA workaround bundled into the GL2 residency PR, and applied to every ISA string

The PR adds a comgr metadata workaround that strips a `-strict` suffix before `get_isa_metadata`. It has nothing to do with GL2 residency control, and the strip is unconditional: any ISA name containing the substring `-strict` anywhere is truncated at that point, not just a trailing `-strict` on gfx1250.

**Suggested fix:** Split this workaround into its own PR so it can be reviewed and reverted independently, and while there, anchor the match (suffix check / explicit gfx1250 guard) and note the comgr issue it works around so it can be removed later.

**Review by mangupta — COMMENTED** [review]
🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** New public limit has no contract test, no uncovered_apis.txt rationale, and no TEST_PLAN.md update

The PR adds a public HIP limit (`hipLimitPersistingL2CacheSize`) with new observable `hipDeviceGetLimit`/`hipDeviceSetLimit` behavior and a changed `hipDeviceProp_t::persistingL2CacheMaxSize`, but adds no `projects/hip-tests/catch/contract/` test with an `@asserts: API - invariant` annotation, no rationale in `projects/hip-tests/catch/contract/uncovered_apis.txt`, and no regeneration of `projects/hip-tests/catch/TEST_PLAN.md`.

**Suggested fix:** Add a `catch/contract/` test asserting the set/get round-trip and the over-max rejection with `@asserts:` annotations (format per `catch/contract/AUTHORING.md`), or record a justified entry in `uncovered_apis.txt`, and regenerate `catch/TEST_PLAN.md`.

File: projects/hip-tests/catch/CMakeLists.txt [hunk 11]
```diff
@@ -3,6 +3,34 @@
...
+  if(EXISTS "${ROCM_PATH}/llvm/bin/${CMAKE_CXX_COMPILER}")
+    set(CMAKE_CXX_COMPILER "${ROCM_PATH}/llvm/bin/${CMAKE_CXX_COMPILER}")
+  endif()
+endif()
+
+if(DEFINED CMAKE_HIP_COMPILER AND NOT IS_ABSOLUTE ${CMAKE_HIP_COMPILER})
+  if(EXISTS "${ROCM_PATH}/llvm/bin/${CMAKE_HIP_COMPILER}")
+    set(CMAKE_HIP_COMPILER "${ROCM_PATH}/llvm/bin/${CMAKE_HIP_COMPILER}")
+  endif()
+endif()
```
Comment (mangupta) [inline projects/hip-tests/catch/CMakeLists.txt]: 🤖 AI-assisted pre-review (for @mangupta)

**[IMPORTANT]** New ROCM_PATH/compiler-path block can hard-error at configure time and sets compiler variables that CMake ignores

Three new `if(DEFINED CMAKE_<LANG>_COMPILER AND NOT IS_ABSOLUTE ${CMAKE_<LANG>_COMPILER})` guards dereference the variable unquoted, and the block then assigns `CMAKE_<LANG>_COMPILER` as a plain (non-cache) variable. Also, this build-environment convenience is unrelated to GL2 residency.

**Suggested fix:** Quote the arguments (`NOT IS_ABSOLUTE "${CMAKE_CXX_COMPILER}"`), and either set the variables as cache entries before `project()`/via a toolchain file or drop the block entirely; consider moving this build-environment convenience to its own PR.
